File: ios/App/App/Studio/Fx.swift
// Native FX chain — maps each FxType in the session schema to an
// AVAudioUnit. Built-in iOS units cover EQ/Reverb/Delay/Distortion.
// Compressor + Filter use AVAudioUnitEffect subclasses (DynamicsProcessor
// at the AU level, lowpass band of AVAudioUnitEQ for filter).
//
// `FxChain.build` chains the nodes in series and returns the head + tail
// the engine wires into the parent mixer.

import Foundation
import AVFoundation

public final class FxChain {
    public let input: AVAudioNode
    public let output: AVAudioNode
    public let nodes: [FxNodeBinding]

    private let engine: AVAudioEngine

    private init(engine: AVAudioEngine, input: AVAudioNode, output: AVAudioNode, nodes: [FxNodeBinding]) {
        self.engine = engine
        self.input = input
        self.output = output
        self.nodes = nodes
    }

    /// Returns nil if specs is empty — caller should just wire input→output directly.
    public static func build(engine: AVAudioEngine, specs: [Studio.FxNode]) -> FxChain? {
        let enabled = specs.filter { $0.enabled }
        guard !enabled.isEmpty else { return nil }

        var bindings: [FxNodeBinding] = []
        for spec in enabled {
            let b = FxNodeBinding.build(engine: engine, spec: spec)
            bindings.append(b)
        }
        // Chain them in series.
        for i in 0..<(bindings.count - 1) {
            engine.connect(bindings[i].node, to: bindings[i + 1].node, format: nil)
        }
        return FxChain(
            engine: engine,
            input: bindings.first!.node,
            output: bindings.last!.node,
            nodes: bindings)
    }

    public func dispose() {
        for n in nodes {
            engine.disconnectNodeInput(n.node)
            engine.detach(n.node)
        }
    }
}

public final class FxNodeBinding {
    public let type: String
    public let node: AVAudioNode

    init(type: String, node: AVAudioNode) {
        self.type = type
        self.node = node
    }

    public static func build(engine: AVAudioEngine, spec: Studio.FxNode) -> FxNodeBinding {
        let node: AVAudioNode
        switch spec.type {
        case .gain:
            // AVAudioEngine has no dedicated gain node; an EQ with all
            // bands flat and a globalGain shift acts as one.
            let eq = AVAudioUnitEQ(numberOfBands: 1)
            eq.globalGain = paramFloat(spec.params, "gain_db", 0)
            node = eq
        case .eq3:
            let eq = AVAudioUnitEQ(numberOfBands: 3)
            eq.bands[0].filterType = .lowShelf
            eq.bands[0].frequency = 200
            eq.bands[0].gain = paramFloat(spec.params, "low_db", 0)
            eq.bands[0].bypass = false
            eq.bands[1].filterType = .parametric
            eq.bands[1].frequency = paramFloat(spec.params, "mid_hz", 1000)
            eq.bands[1].bandwidth = 1.0
            eq.bands[1].gain = paramFloat(spec.params, "mid_db", 0)
            eq.bands[1].bypass = false
            eq.bands[2].filterType = .highShelf
            eq.bands[2].frequency = 5000
            eq.bands[2].gain = paramFloat(spec.params, "high_db", 0)
            eq.bands[2].bypass = false
            node = eq
        case .compressor:
            // AVAudioUnitDynamicsProcessor is the high-level wrapper.
            let dyn = AVAudioUnitEffect(audioComponentDescription: .init(
                componentType: kAudioUnitType_Effect,
                componentSubType: kAudioUnitSubType_DynamicsProcessor,
                componentManufacturer: kAudioUnitManufacturer_Apple,
                componentFlags: 0, componentFlagsMask: 0))
            // Param setting via AudioUnit indices — see Apple's DynamicsProcessor IDs.
            // 0 Threshold (dB), 1 HeadRoom, 2 ExpansionRatio, 3 ExpansionThreshold,
            // 4 AttackTime, 5 ReleaseTime, 6 MasterGain.
            AudioUnitSetParameter(dyn.audioUnit, 0, kAudioUnitScope_Global, 0,
                                  paramFloat(spec.params, "threshold_db", -18), 0)
            AudioUnitSetParameter(dyn.audioUnit, 6, kAudioUnitScope_Global, 0,
                                  paramFloat(spec.params, "makeup_db", 0), 0)
            // Attack/release in seconds.
            AudioUnitSetParameter(dyn.audioUnit, 4, kAudioUnitScope_Global, 0,
                                  paramFloat(spec.params, "attack_ms", 5) / 1000.0, 0)
            AudioUnitSetParameter(dyn.audioUnit, 5, kAudioUnitScope_Global, 0,
                                  paramFloat(spec.params, "release_ms", 80) / 1000.0, 0)
            node = dyn
        case .reverb:
            let rev = AVAudioUnitReverb()
            rev.loadFactoryPreset(.mediumHall)
            rev.wetDryMix = paramFloat(spec.params, "wet", 0.25) * 100
            node = rev
        case .delay:
            let del = AVAudioUnitDelay()
            del.delayTime = TimeInterval(paramFloat(spec.params, "time_ms", 350) / 1000.0)
            del.feedback = paramFloat(spec.params, "feedback", 0.35) * 100
            del.wetDryMix = paramFloat(spec.params, "wet", 0.25) * 100
            node = del
        case .filter:
            // A single-band EQ acts as a high/low/band filter.
            let eq = AVAudioUnitEQ(numberOfBands: 1)
            let band = eq.bands[0]
            let kind = paramString(spec.params, "kind", "low")
            switch kind {
            case "low":  band.filterType = .lowPass
            case "high": band.filterType = .highPass
            case "band": band.filterType = .bandPass
            default:     band.filterType = .lowPass
            }
            band.frequency = paramFloat(spec.params, "cutoff_hz", 1000)
            band.bandwidth = max(0.1, paramFloat(spec.params, "q", 0.7))
            band.bypass = false
            node = eq
        }
        engine.attach(node)
        return FxNodeBinding(type: spec.type.rawValue, node: node)
    }
}

// ── Param accessors ──────────────────────────────────────────────────

func paramFloat(_ p: [String: Studio.ParamValue], _ key: String, _ fallback: Float) -> Float {
    guard case .number(let v) = p[key] else { return fallback }
    return Float(v)
}
func paramString(_ p: [String: Studio.ParamValue], _ key: String, _ fallback: String) -> String {
    guard case .string(let v) = p[key] else { return fallback }
    return v
}
func paramBool(_ p: [String: Studio.ParamValue], _ key: String, _ fallback: Bool) -> Bool {
    guard case .bool(let v) = p[key] else { return fallback }
    return v
}

File: src/lib/studio/engine/fx.ts
// Map each FxType in the session schema to a concrete Tone.js node.
// `buildFxChain` returns an object exposing one input + one output —
// the engine wires those into the master bus or a track strip.
//
// Param keys here MUST match the defaults declared in defaults.ts so
// presets round-trip through save/load.

import * as Tone from 'tone';
import type { FxNode } from '../session';

export interface EngineFxNode {
  type: FxNode['type'];
  inputOutput: Tone.ToneAudioNode;   // node that's both input & output
  applyParams: (params: FxNode['params']) => void;
  setEnabled: (enabled: boolean) => void;
  dispose: () => void;
}

export interface EngineFxChain {
  input: Tone.ToneAudioNode;
  output: Tone.ToneAudioNode;
  nodes: EngineFxNode[];
  dispose: () => void;
}

function num(v: unknown, fallback: number): number {
  return typeof v === 'number' && Number.isFinite(v) ? v : fallback;
}

export function buildFxNode(spec: FxNode): EngineFxNode {
  switch (spec.type) {
    case 'gain': {
      const node = new Tone.Gain(num(spec.params.gain_db, 0) === 0
        ? 1 : Math.pow(10, num(spec.params.gain_db, 0) / 20));
      return {
        type: 'gain', inputOutput: node,
        applyParams: (p) => { node.gain.value = Math.pow(10, num(p.gain_db, 0) / 20); },
        setEnabled: () => { /* gain handled via bypass node */ },
        dispose: () => node.dispose(),
      };
    }
    case 'eq3': {
      const node = new Tone.EQ3(num(spec.params.low_db, 0), num(spec.params.mid_db, 0), num(spec.params.high_db, 0));
      return {
        type: 'eq3', inputOutput: node,
        applyParams: (p) => {
          node.low.value = num(p.low_db, 0);
          node.mid.value = num(p.mid_db, 0);
          node.high.value = num(p.high_db, 0);
        },
        setEnabled: () => {},
        dispose: () => node.dispose(),
      };
    }
    case 'compressor': {
      const node = new Tone.Compressor({
        threshold: num(spec.params.threshold_db, -18),
        ratio: num(spec.params.ratio, 3),
        attack: num(spec.params.attack_ms, 5) / 1000,
        release: num(spec.params.release_ms, 80) / 1000,
      });
      return {
        type: 'compressor', inputOutput: node,
        applyParams: (p) => {
          node.threshold.value = num(p.threshold_db, -18);
          node.ratio.value = num(p.ratio, 3);
          node.attack.value = num(p.attack_ms, 5) / 1000;
          node.release.value = num(p.release_ms, 80) / 1000;
        },
        setEnabled: () => {},
        dispose: () => node.dispose(),
      };
    }
    case 'reverb': {
      const node = new Tone.Reverb({
        wet: num(spec.params.wet, 0.25),
        decay: 1 + num(spec.params.room_size, 0.6) * 3,
      });
      return {
        type: 'reverb', inputOutput: node,
        applyParams: (p) => {
          node.wet.value = num(p.wet, 0.25);
          // decay change requires async regenerate — keep simple here.
        },
        setEnabled: () => {},
        dispose: () => node.dispose(),
      };
    }
    case 'delay': {
      const node = new Tone.FeedbackDelay({
        delayTime: num(spec.params.time_ms, 350) / 1000,
        feedback: num(spec.params.feedback, 0.35),
        wet: num(spec.params.wet, 0.25),
      });
      return {
        type: 'delay', inputOutput: node,
        applyParams: (p) => {
          node.delayTime.value = num(p.time_ms, 350) / 1000;
          node.feedback.value = num(p.feedback, 0.35);
          node.wet.value = num(p.wet, 0.25);
        },
        setEnabled: () => {},
        dispose: () => node.dispose(),
      };
    }
    case 'filter': {
      const kind = (spec.params.kind as string) ?? 'low';
      const map: Record<string, 'lowpass' | 'highpass' | 'bandpass'> = {
        low: 'lowpass', high: 'highpass', band: 'bandpass',
      };
      const node = new Tone.Filter({
        type: map[kind] ?? 'lowpass',
        frequency: num(spec.params.cutoff_hz, 1000),
        Q: num(spec.params.q, 0.7),
      });
      return {
        type: 'filter', inputOutput: node,
        applyParams: (p) => {
          const k = (p.kind as string) ?? 'low';
          node.type = (map[k] ?? 'lowpass');
          node.frequency.value = num(p.cutoff_hz, 1000);
          node.Q.value = num(p.q, 0.7);
        },
        setEnabled: () => {},
        dispose: () => node.dispose(),
      };
    }
  }
}

/** Chain a list of FX in series. Bypassed nodes are skipped at build
 * time — toggling enabled in the UI rebuilds the chain. */
export function buildFxChain(specs: FxNode[]): EngineFxChain {
  const nodes = specs.filter((s) => s.enabled).map(buildFxNode);
  // A pair of gain nodes flank the chain so the engine has stable
  // input/output references even when nodes change.
  const input = new Tone.Gain(1);
  const output = new Tone.Gain(1);

  if (nodes.length === 0) {
    input.connect(output);
  } else {
    input.connect(nodes[0].inputOutput);
    for (let i = 0; i < nodes.length - 1; i++) {
      nodes[i].inputOutput.connect(nodes[i + 1].inputOutput);
    }
    nodes[nodes.length - 1].inputOutput.connect(output);
  }

  return {
    input, output, nodes,
    dispose: () => {
      input.dispose();
      output.dispose();
      for (const n of nodes) n.dispose();
    },
  };
}
